pkg/consensus/consensus.go

```diff
@@ -679,7 +679,7 @@ func (s *service) getValidators(txes ...block.Transaction) []crypto.PublicKey {
 	if txes == nil {
```

Comment:
BTW, this likely doesn't make much sense now. IIRC it was an optimization. Needs to be checked wrt C# though.

Reply:
Not exactly. 7eef8950612b37f83da2f3639d32fd61800a6670

Reply:
IMHO, it's definitely not the optimisation. See the 7eef8950, and I've checked the C# node: we use Blockchain's `GetNextBlockValidators` API to retrieve the list of validators that is constant and is updated once per N blocks (N - committee size). And we use `getValidators()` without parameters to initialize the set of validators who will sign the subsequent dBFT block. This part matches the C#.

And we use `getValidators(txs...)` to retrieve the `NextConsensus` field for the subsequent dBFT block. And according to our code, NextConsensus address can be changed at every block, because it fetches the most fresh candidates/votes data and calculates committee based on it. See the updated methods comments. And now I'm doubting that this part matches the C#, let's discuss it on the meeting tomorrow.

Reply:
```
During the consensus phase, a speaker will send PrepareRequest message with NextConsensus , which determines the next block consensus nodes. The Speaker gets the next round of consensus nodes by combining the transactions in the proposal block with the previous votes in blockchain, and assign the script hash of 2/3 multi-signature contract to NextConsensus .
```

Reply:
The doc is just not correct. This is the way it was on Legacy:
https://github.com/nspcc-dev/neo-go/blob/fae1d562e7852e57c9f2077c6fc913c3e955204b/pkg/consensus/consensus.go#L625-L633
Where votes were really processed for the new set of transactions:
https://github.com/nspcc-dev/neo-go/blob/fae1d562e7852e57c9f2077c6fc913c3e955204b/pkg/core/blockchain.go#L2159-L2164
On N3 that would correspond to executing the whole set of transactions right in the consensus process, but we don't do it, at block N we build N+1 using validators from the state N-1 (to sign N+1) and including validators from the current state N into the `NextConsensus`. The set of transactions of N+1 will influence the `NextConsensus` at N+2 and will be used for N+3.

But if we have (cached, committee-based) `GetNextBlockValidators()` _always_ used to determine the _current_ set of validators (while picking the NextConsensus ones from dynamic voting), we might have a problem changing the set if there are more than F nodes changed. This needs to be checked against C#.